services/mlmodel/tflite_cpu/tflite_cpu.go

```diff
@@ -0,0 +1,233 @@
+	}
+	return CreateTFLiteCPUModel(ctx, params)
+}
+
+// CreateTFLiteCPUModel is a constructor that builds a tflite cpu implementation of the MLMS.
+func CreateTFLiteCPUModel(ctx context.Context, params *TFLiteConfig) (*Model, error) {
+	// Given those params, add the model
+	model, err := addTFLiteModel(ctx, params.ModelPath, &params.NumThreads)
+	if err != nil {
+		return &Model{}, errors.Wrapf(err, "could not add model from location %s", params.ModelPath)
+	}
+	return &Model{attrs: *params, model: model}, nil
+}
+
+// Infer takes the input map, finds the image (labeled 'image'), and uses the inference
+// package to return the result from the tflite cpu model as a map.
+func (m *Model) Infer(ctx context.Context, input map[string]interface{}) (map[string]interface{}, error) {
+	_, span := trace.StartSpan(ctx, "service::mlmodel::tflite_cpu::Infer")
+	defer span.End()
+
+	outMap := make(map[string]interface{})
+
+	// Grab the image data from the map
+	if imgBytes, ok := input["image"]; ok {
+		outTensors, err := m.model.Infer(imgBytes)
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't infer from model")
+		}
+		// Fill in the output map with the names from metadata if u have them
+		// Otherwise, do output1, output2, etc.
+		for i := 0; i < len(m.metadata.Outputs); i++ {
+			name := m.metadata.Outputs[i].Name
+			if name == "" {
+				outMap["output"+strconv.Itoa(i)] = outTensors[i]
+			} else {
+				outMap[m.metadata.Outputs[i].Name] = outTensors[i]
+			}
+		}
+	} else {
+		return map[string]interface{}{}, errors.New("could not find image in input map. Give it the name 'image' duh")
+	}
+
+	return outMap, nil
+}
+
+// Metadata reads the metadata from your tflite cpu model into the metadata struct
+// that we use for the mlmodel service.
+func (m *Model) Metadata(ctx context.Context) (mlmodel.MLMetadata, error) {
+	_, span := trace.StartSpan(ctx, "service::mlmodel::tflite_cpu::Metadata")
+	defer span.End()
+
+	// model.Metadata() and funnel it into this struct
+	md, err := m.model.Metadata()
+	if err != nil {
+		return mlmodel.MLMetadata{}, err
+	}
+	out := mlmodel.MLMetadata{}
+	out.ModelName = md.Name
+	out.ModelDescription = md.Description
+
+	numIn := m.model.Info.InputTensorCount
+	numOut := m.model.Info.OutputTensorCount
+	inputList := make([]mlmodel.TensorInfo, 0, numIn)
+	outputList := make([]mlmodel.TensorInfo, 0, numOut)
+
+	// Fill in input info to the best of our abilities (normally just 1 tensor)
+	for i := 0; i < numIn; i++ { // for each input Tensor
+		inputT := md.SubgraphMetadata[0].InputTensorMetadata[i]
+		td := getTensorInfo(inputT)
+		// try to guess model type based on input description
+		if strings.Contains(inputT.Description, "detect") {
+			out.ModelType = "tflite_detector"
+		}
+		if strings.Contains(inputT.Description, "classif") {
+			out.ModelType = "tflite_classifier"
+		}
+
+		switch m.model.Info.InputTensorType { // grab from model info, not metadata
```

Comment:
Is this switch statement exhaustive? Are there more data types that are possible from the model info?

Reply:
Yup it is (see ml/inference/tflite.go).  Although I suspect this list was created just to handle what we've seen.